cardano-cli/src/Cardano/CLI/Ping/Lib.hs

```diff
@@ -0,0 +1,463 @@
+
+data HandshakeFailure
+  = UnknownVersionInRsp !Word
+  | UnknownKey !Word
+  | UnknownTag !Word
+  | VersionMissmath ![Word]
+  | DecodeError !Word !String
+  | Refused !Word !String
+  deriving Show
+
+newtype KeepAliveFailure = KeepAliveFailureKey Word deriving Show
+
+keepAliveRspDec :: NodeVersion
+                -> CBOR.Decoder s (Either KeepAliveFailure Word16)
+keepAliveRspDec v | v >= NodeToNodeVersionV7 minBound minBound = do
+  len <- CBOR.decodeListLen
+  key <- CBOR.decodeWord
+  case (len, key) of
+    (2, 1) -> Right <$> CBOR.decodeWord16
+    (_, k) -> return $ Left $ KeepAliveFailureKey k
+keepAliveRspDec _ = do
+  key <- CBOR.decodeWord
+  case key of
+    1 -> Right <$> CBOR.decodeWord16
+    k -> return $ Left $ KeepAliveFailureKey k
+
+handshakeDec :: CBOR.Decoder s (Either HandshakeFailure NodeVersion)
+handshakeDec = do
+  _ <- CBOR.decodeListLen
+  key <- CBOR.decodeWord
+  case key of
+    1 -> decodeVersion
+    2 -> do
+      _ <- CBOR.decodeListLen
+      tag <- CBOR.decodeWord
+      case tag of
+        0 -> do -- VersionMismatch
+          len <- CBOR.decodeListLen
+          x <- replicateM len CBOR.decodeWord
+          return $ Left $ VersionMissmath x
+        1 -> do -- HandshakeDecodeError
+          vn <- CBOR.decodeWord
+          msg <- unpack <$> CBOR.decodeString
+          return $ Left $ DecodeError vn msg
+        2 -> do -- Refused
+          vn <- CBOR.decodeWord
+          msg <- unpack <$> CBOR.decodeString
+          return $ Left $ Refused vn msg
+        _ -> return $ Left $ UnknownTag tag
+
+    k -> return $ Left $ UnknownKey k
+  where
+    decodeVersion :: CBOR.Decoder s (Either HandshakeFailure NodeVersion)
+    decodeVersion = do
+      version <- CBOR.decodeWord
+      let cb = version `clearBit` nodeToClientVersionBit
+      let tb = version `testBit`  nodeToClientVersionBit
+      case (cb, tb) of
+        (7,  False) -> decodeWithMode NodeToNodeVersionV7
+        (8,  False) -> decodeWithMode NodeToNodeVersionV8
+        (9,  False) -> decodeWithMode NodeToNodeVersionV9
+        (10, False) -> decodeWithMode NodeToNodeVersionV10
+        (9,  True)  -> Right . NodeToClientVersionV9 <$> CBOR.decodeWord32
+        (10, True)  -> Right . NodeToClientVersionV10 <$> CBOR.decodeWord32
+        (11, True)  -> Right . NodeToClientVersionV11 <$> CBOR.decodeWord32
+        (12, True)  -> Right . NodeToClientVersionV12 <$> CBOR.decodeWord32
+        (13, True)  -> Right . NodeToClientVersionV13 <$> CBOR.decodeWord32
+        (14, True)  -> Right . NodeToClientVersionV14 <$> CBOR.decodeWord32
+        _           -> return $ Left $ UnknownVersionInRsp version
+
+    decodeWithMode :: (Word32 -> Bool -> NodeVersion) -> CBOR.Decoder s (Either HandshakeFailure NodeVersion)
+    decodeWithMode vnFun = do
+      _ <- CBOR.decodeListLen
+      magic <- CBOR.decodeWord32
+      Right . vnFun magic <$> CBOR.decodeBool
+
+wrap :: MiniProtocolNum -> MiniProtocolDir -> LBS.ByteString -> MuxSDU
```

Comment:
A comment here explaining this function, its type parameters and return type would be useful

Reply:
This has moved to `ouroboros-network`.